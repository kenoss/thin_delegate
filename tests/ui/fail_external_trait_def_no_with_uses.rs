// Compare with fail_external_trait_def_with_uses.rs
// See also //tests/ui/smithay/*.rs

mod external {
    pub struct Arg;

    pub trait Hello {
        fn hello(&self, arg: Arg) -> String;
    }

    pub trait Hi {
        fn hi(&self, arg: Arg) -> String;
    }

    impl Hello for String {
        fn hello(&self, _arg: Arg) -> String {
            format!("hello, {self}")
        }
    }

    impl Hi for String {
        fn hi(&self, _arg: Arg) -> String {
            format!("hi, {self}")
        }
    }
}

#[thin_delegate::external_trait_def]
mod __external_trait_def {
    // Meaningless. `#[thin_delegate::register]` registers a token stream as is.
    use crate::external::Arg;

    #[thin_delegate::register]
    pub trait Hello {
        // OK. Replaced `Arg` with full path.
        fn hello(&self, arg: crate::external::Arg) -> String;
    }

    #[thin_delegate::register]
    pub trait Hi {
        // Not good. In `#[thin_delegate::fill_delegate]`, it will be expanded as is. So, one
        // needs to `use` for each derive. It's not convernient if a trait uses lots of types.
        fn hi(&self, arg: Arg) -> String;
    }
}

#[thin_delegate::register]
struct Hoge(String);

// OK.
#[thin_delegate::fill_delegate(external_trait_def = __external_trait_def)]
impl external::Hello for Hoge {}

// NG.
#[thin_delegate::fill_delegate(external_trait_def = __external_trait_def)]
impl external::Hi for Hoge {}

fn main() {
    use external::{Arg, Hello, Hi};

    let hoge = Hoge("world".to_string());
    assert_eq!(hoge.hello(Arg), "hello, world");
    assert_eq!(hoge.hi(Arg), "hi, world");
}
