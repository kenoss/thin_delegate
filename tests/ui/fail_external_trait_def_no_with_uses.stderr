error[E0412]: cannot find type `Arg` in this scope
  --> tests/ui/fail_external_trait_def_no_with_uses.rs:43:27
   |
43 |         fn hi(&self, arg: Arg) -> String;
   |                           ^^^ not found in this scope
...
55 | #[thin_delegate::fill_delegate(external_trait_def = __external_trait_def)]
   | -------------------------------------------------------------------------- in this procedural macro expansion
   |
   = note: this error originates in the macro `__external_trait_def::__thin_delegate__feed_trait_def_of_external_Hi` which comes from the expansion of the attribute macro `thin_delegate::fill_delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing this struct
   |
4  + use crate::external::Arg;
   |

warning: unused import: `crate::external::Arg`
  --> tests/ui/fail_external_trait_def_no_with_uses.rs:31:9
   |
31 |     use crate::external::Arg;
   |         ^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
